src/rust/graph-descriptions/src/dynamic_node.rs

```diff
@@ -123,4 +98,60 @@ impl NodeT for DynamicNode {
 
         merged
     }
+
+    fn attach_predicates_to_mutation_unit(&self, mutation_unit: &mut MutationUnit) {
+        mutation_unit.predicate_ref("node_key", MutationPredicateValue::string(&self.node_key));
+        mutation_unit.predicate_ref(
+            "seen_at",
+            MutationPredicateValue::Number(self.seen_at as i64),
+        );
+        mutation_unit.predicate_ref(
+            "dgraph.type",
+            MutationPredicateValue::string(&self.node_type),
+        );
+
+        if let Some(asset_id) = &self.asset_id {
+            mutation_unit.predicate_ref("asset_id", MutationPredicateValue::string(asset_id));
+        }
+
+        for (key, prop) in &self.properties {
+            let prop = match &prop.property {
+                Some(node_property::Property::Intprop(i)) => {
+                    MutationPredicateValue::Number(*i as i64)
+                }
+                Some(node_property::Property::Uintprop(i)) => {
+                    MutationPredicateValue::Number(*i as i64)
+                }
+                Some(node_property::Property::Strprop(s)) => MutationPredicateValue::string(s),
+                None => panic!("Invalid property on DynamicNode: {}", self.node_key),
+            };
+
+            mutation_unit.predicate_ref(key, prop);
+        }
+    }
+
+    fn get_cache_identities_for_predicates(&self) -> Vec<Vec<u8>> {
+        let mut predicate_cache_identities = Vec::new();
```

Comment:
Not gonna block on it, and idc if you fix it at all, but there's `Vec::with_capacity`, which is good for cases where you know how large the Vec will get to, or at least a lower bound. So here I expect it would be `Vec::with_capacity(self.properties.len())`

Again, don't care, just a nice quick win as you write code.